Shape_detection/include/CGAL/Shape_detection/Efficient_RANSAC/Octree.h

```diff
@@ -26,699 +26,706 @@
+  }
+
+  std::size_t offset() {
+    return m_offset;
+  }
+
+  std::size_t size() {
+    return m_beyond - m_first + 1;
+  }
+
+  Input_iterator first() {
+    return m_first;
+  }
+
+  Input_iterator beyond() {
+    return m_beyond;
   }
+
+  void setData(Input_iterator &begin, Input_iterator &beyond) {
+    m_beyond = (beyond == begin) ? begin : beyond - 1;
+  }
+
+  void swap(std::size_t a, std::size_t b) {
+    typename std::iterator_traits<Input_iterator>::value_type tmp;
+    tmp = m_first[a];
+    m_first[a] = m_first[b];
+    m_first[b] = tmp;
+  }
+
+protected:
+  Input_iterator m_first;
+
+private:
+  Input_iterator m_beyond;
+  std::size_t m_offset;
+};
+
+template<class Sdt>
+class IndexedPointAccessor {
+public:
+  typedef Sdt Sd_traits;
+  typedef typename Sd_traits::Input_range::iterator Input_iterator;
+
+  IndexedPointAccessor() {}
+
+  IndexedPointAccessor(const Input_iterator &begin,
+                       const Input_iterator &beyond, std::size_t)
+          : m_first(begin) {
+    m_beyond = (beyond == begin) ? begin : beyond - 1;
+    m_indices.resize(size());
+    for (std::size_t i = 0; i < size(); i++)
+      m_indices[i] = i;
+  }
+
+  Input_iterator at(std::size_t i) {
+    return m_first + m_indices[i];
+  }
+
+  std::size_t index(std::size_t i) {
+    return m_indices[i];
+  }
+
+  std::size_t offset() {
+    return 0;
+  }
+
+  Input_iterator first() {
+    return m_first;
+  }
+
+  Input_iterator beyond() {
+    return m_beyond;
+  }
+
+  void setData(Input_iterator &begin, Input_iterator &beyond) {
+    m_beyond = (beyond == begin) ? begin : beyond - 1;
+    m_indices.resize(size());
+    for (std::size_t i = 0; i < size(); i++)
+      m_indices[i] = i;
   }
 
+  std::size_t size() {
+    return m_beyond - m_first + 1;
+  }
+
+  void swap(std::size_t a, std::size_t b) {
+    std::size_t tmp = m_indices[a];
+    m_indices[a] = m_indices[b];
+    m_indices[b] = tmp;
+  }
+
+protected:
+  Input_iterator m_first;
+
+private:
+  std::vector<std::size_t> m_indices;
+  Input_iterator m_beyond;
+};
+
+template<class PointAccessor>
+class Octree : public PointAccessor {
+
+  typedef typename PointAccessor::Sd_traits Sd_traits;
+  typedef typename Sd_traits::Input_range::iterator Input_iterator;
+  typedef Shape_base<Sd_traits> Shape;
+  typedef typename Sd_traits::Point_3 Point_3;
+  typedef typename Sd_traits::Vector_3 Vector_3;
+  typedef typename Sd_traits::FT FT;
+  typedef typename Sd_traits::Point_map Point_map;
+  typedef typename Sd_traits::Normal_map Normal_map;
+
+  template<class Sd_traits>
+  friend
+  class ::CGAL::Shape_detection::Efficient_RANSAC;
+
+//      struct Cell {
+//        std::size_t first, last;
+//        Cell *child[8];
+//        Point_3 center;
+//        std::size_t level;
+//
+//        Cell(std::size_t first, std::size_t last, Point_3 center, std::size_t level)
+//          : first(first), last(last), center(center), level(level) {
+//            memset(child, 0, sizeof(Cell *) * 8);
```

Comment:
Please do not comment lines to "remove" code. Just remove the lines. We have the Git history to retrieve what has disappeared.

Reply:
Thanks for the advice! After some experimenting, it looks like turning the RANSAC tree into a wrapper won't be a workable intermediate step after all. It exposes its member variables, and there's a lot of code that interacts with them directly.

I have my meeting with Simon tomorrow, and we should be able to find a better way forward! For now I removed the old octree entirely, and I'm spending some time trying to get an understanding for the way it was used.